File: homework3/go/exercises_test.go
package exercises

import (
        "fmt"
)

func ExampleChange() {
        cases := []int{5, 25, 97, 1000000000, -30}
        for _, amount := range cases {
                coins, err := Change(amount)
                if err != nil {
                        fmt.Println(err)
                } else {
                        fmt.Println(coins)
                }
        }
        // Output: [0 0 1 0]
        // [1 0 0 0]
        // [3 2 0 2]
        // [40000000 0 0 0]
        // amount cannot be negative
}

func ExamplePowers() {
        ch := make(chan int, 100)
        go Powers(2, -5, ch)
        go Powers(7, 0, ch)
        go Powers(3, 1, ch)
        fmt.Print(<-ch)
        go Powers(2, 63, ch)
        for i := 0; i < 6; i++ {
                fmt.Print(<-ch)
        }
        go Powers(2, 64, ch)
        for i := 0; i < 7; i++ {
                fmt.Print(<-ch)
        }
        fmt.Println()
        // Output: 1124816321248163264
}

func ExampleCylinder() {
        c := Cylinder{3.0, 5.0}
        fmt.Printf("%9.5f\n", c.SurfaceArea())
        fmt.Printf("%9.5f\n", c.Volume())
        // Output: 150.79645
        // 141.37167
}

File: homework3/erlang/exercises_test.erl
%
expectReceive(N) ->
    receive P -> P = N end.

main(_) ->
    {true, [0, 0, 1, 3]} = exercises:change(8),
    {true, [3, 2, 0, 2]} = exercises:change(97),
    {true, [4000000, 0, 0, 1]} = exercises:change(100000001),
    {true, [1, 1, 1, 2]} = exercises:change(42),
    {false, "amount cannot be negative"} = exercises:change(-5),

    219 = exercises:sumOfCubesOfOdds([7, 1, 0, 10, -5]),
    0 = exercises:sumOfCubesOfOdds([0, 2, 10000]),
    0 = exercises:sumOfCubesOfOdds([]),
    -1 = exercises:sumOfCubesOfOdds([0, -1, 8]),

    Main = self(),
    spawn(exercises, powers, [2, -5, Main]),
    spawn(exercises, powers, [7, 0, Main]),
    spawn(exercises, powers, [3, 1, Main]),
    expectReceive(1),
    spawn(exercises, powers, [2, 63, Main]),
    [expectReceive(N) || N <- [1, 2, 4, 8, 16, 32]],
    spawn(exercises, powers, [2, 64, Main]),
    [expectReceive(N) || N <- [1, 2, 4, 8, 16, 32, 64]],
    io:format("All tests passed~n", []).